heron/ckptmgr/src/java/com/twitter/heron/ckptmgr/CheckpointManagerServer.java

```diff
@@ -0,0 +1,298 @@
+      IStatefulStorage checkpointsBackend, NIOLooper looper, String host,
+      int port, HeronSocketOptions options) {
+    super(looper, host, port, options);
+
+    this.topologyName = topologyName;
+    this.topologyId = topologyId;
+    this.checkpointMgrId = checkpointMgrId;
+    this.checkpointsBackend = checkpointsBackend;
+
+    this.connection = null;
+
+    registerInitialization();
+  }
+
+  private void registerInitialization() {
+    registerOnRequest(CheckpointManager.RegisterStMgrRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.RegisterTMasterRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.SaveInstanceStateRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.GetInstanceStateRequest.newBuilder());
+
+    registerOnRequest(CheckpointManager.CleanStatefulCheckpointRequest.newBuilder());
+  }
+
+  @Override
+  public void onConnect(SocketChannel channel) {
+    LOG.info("Got a new connection from host:port "
+        + channel.socket().getRemoteSocketAddress());
+  }
+
+  @Override
+  public void onRequest(REQID rid, SocketChannel channel, Message request) {
+    if (request instanceof CheckpointManager.RegisterStMgrRequest) {
+      handleStMgrRegisterRequest(rid, channel, (CheckpointManager.RegisterStMgrRequest) request);
+    } else if (request instanceof CheckpointManager.RegisterTMasterRequest) {
+      handleTMasterRegisterRequest(rid, channel,
+                                   (CheckpointManager.RegisterTMasterRequest) request);
+    } else if (request instanceof CheckpointManager.SaveInstanceStateRequest) {
+      handleSaveInstanceStateRequest(
+          rid, channel, (CheckpointManager.SaveInstanceStateRequest) request);
+    } else if (request instanceof CheckpointManager.GetInstanceStateRequest) {
+      handleGetInstanceStateRequest(
+          rid, channel, (CheckpointManager.GetInstanceStateRequest) request);
+    } else if (request instanceof CheckpointManager.CleanStatefulCheckpointRequest) {
+      handleCleanStatefulCheckpointRequest(
+          rid, channel, (CheckpointManager.CleanStatefulCheckpointRequest) request);
+    } else {
+      LOG.severe("Unknown kind of request: " + request.getClass().getName());
+    }
+  }
+
+  protected void handleCleanStatefulCheckpointRequest(
+      REQID rid,
+      SocketChannel channel,
+      CheckpointManager.CleanStatefulCheckpointRequest request
+  ) {
+    LOG.info("Got a clean request from " + request.toString() + " host:port "
+        + channel.socket().getRemoteSocketAddress());
+
+    boolean deleteAll = request.hasCleanAllCheckpoints() && request.getCleanAllCheckpoints();
+    Common.StatusCode statusCode = Common.StatusCode.OK;
+
+    try {
+      checkpointsBackend.dispose(topologyName,
+                                 request.getOldestCheckpointPreserved(), deleteAll);
+      LOG.info("Dispose checkpoint successful");
+    } catch (StatefulStorageException e) {
```

Comment:
That's totally sufficient for the implementors of IStorageInterface. The issue I'm seeing is that one of them throws an unchecked exception due to a bug. In that case that exception seems like it will get thrown all the way up to HeronServer.handleAccept and then continue up to the slave looper Runnable, which would kill the looper thread and leave the server in a zombie state. Somewhere `RuntimeException` needs to be handled to avoid this possibility.

Ideally we could do that generically in HeronServer, but since it doesn't know what type of response to return in that situation it can't handle that. If we had a generic interface to handle different requests/responses though, we could decorate that in a single generic handler that catches RuntimeException and returns the appropriate response, as defined by the implementor of the handler interface. Or maybe HeronServer handle it generically if it was extended to have a map of response types, just like it has a map of request types...